### ISSUE TITLE
[YSQL] Feature support - ALTER TABLE

### ISSUE THREAD

**Issue by nocaway** [YugaByte/yugabyte-db#1124]
All of ALTER TABLE commands are not supported except for the following.
- ADD <column> without any constraints.
- DROP <column> that is not part of a PRIMARY KEY
- ADD table constraints
- DROP table constraints
- RENAME table

**Comment by dodilp** [comment]
Also fails to work with queries like below -
ALTER TABLE ONLY bookings
    ADD CONSTRAINT fk_bookings_facid FOREIGN KEY (facid) REFERENCES facilities(facid);

**Comment by kmuthukk** [comment]
@dodilp - FOREIGN KEY is being implemented as we speak. Hope to have it in a few weeks.

**Comment by ramanans** [comment]
Fails for 
ALTER TABLE public.<table_name> OWNER TO postgres; 

ALTER TABLE OWNER not supported yet

**Comment by igremmerlb** [comment]
For us we are getting the following running migrations from Entity Framework:

Failed executing DbCommand (10ms) [Parameters=[], CommandType='Text', CommandTimeout='30']
ALTER TABLE "KeySetRoleAssignment" ALTER COLUMN "PermissionId" TYPE varchar(128);
ALTER TABLE "KeySetRoleAssignment" ALTER COLUMN "PermissionId" SET NOT NULL;
ALTER TABLE "KeySetRoleAssignment" ALTER COLUMN "PermissionId" DROP DEFAULT;
Npgsql.PostgresException (0x80004005): 0A000: ALTER TABLE ALTER column not supported yet

**Comment by fire** [comment]
Fails to be used as https://www.metabase.com/'s relational database backend.

**Comment by rkarthik007** [comment]
cc @nocaway @m-iancu : any thoughts? Maybe we don't yet support these alter table variants?

**Comment by ddorian** [comment]
Also missing `ALTER TABLE RENAME CONSTRAINT`. Used in kong migrations #2829

**Comment by irizzant** [comment]
+1 For this issue cause I can't restore my dump for this

**Comment by ddorian** [comment]
@irizzant which exact statement is blocking you ?

**Comment by irizzant** [comment]
@ddorian 
```
pg_restore: [archiver (db)] could not execute query: ERROR:  ALTER TABLE ALTER column not supported yet
LINE 14: ALTER TABLE ONLY crm.customer_relationships ALTER COLUMN ind...
                                                     ^
HINT:  See https://github.com/YugaByte/yugabyte-db/issues/1124. Click '+' on the description to raise its priority
    Command was: CREATE TABLE crm.customer_relationships (
    customer_id character(16) NOT NULL,
    index integer NOT NULL,
    relationship_type character varying(20) NOT NULL,
    assigned_by integer NOT NULL,
    assigned_on timestamp without time zone NOT NULL,
    since date NOT NULL,
    upto date,
    deleted_by integer,
    deleted_on timestamp without time zone,
    record_status character(1) NOT NULL
)
WITH (autovacuum_enabled='true');
ALTER TABLE ONLY crm.customer_relationships ALTER COLUMN index SET STATISTICS 0;
```

**Comment by ddorian** [comment]
@irizzant I believe you need this issue then : https://github.com/yugabyte/yugabyte-db/issues/1013 ?

**Comment by irizzant** [comment]
@ddorian ok thank you, I've also subscribed to #1013

**Comment by ddorian** [comment]
@irizzant I don't think we use index-statistics like in postgresql. Can you comment that statement and continue migration and see if there's anything else blocking you ?

**Comment by irizzant** [comment]
@ddorian not an easy task, I have a whole company db dump I have to try and search/replace for occurrences is not feasible.
The blocking issues I'm having, apart from this, are:
* https://github.com/yugabyte/yugabyte-db/issues/3512
* https://github.com/yugabyte/yugabyte-db/issues/1127
* https://github.com/yugabyte/yugabyte-db/issues/2711

**Comment by z0mb1ek** [comment]
+1

**Comment by sergeyshaykhullin** [comment]
+1

**Comment by uwejan** [comment]
Good work guys, i can not migrate my schema for the following reason, is there a way around this?
`*** There was an issue. Reason: This ALTER TABLE command is not yet supported.. Table: null Column: null`

**Comment by frozenspider** [comment]
Hey @uwejan, would you mind posting the exact command you're executing and YB version you're using on [our community slack](https://www.yugabyte.com/slack) or [forum](https://forum.yugabyte.com/) so we could give you more detailed answer?

**Comment by uwejan** [comment]
@frozenspider Hi, i have posted on slack. Thank you.

**Comment by dbotwinick** [comment]
+1 for support for: "ALTER TABLE name CLUSTER ON"

I'm looking to use yugabytedb as a stand-in for pgsql for an existing application. I'm looking to see if I can modify the application & schema as needed, but obviously it'd be easier for me to use yugabytedb as a 'drop-in' replacement if it did not throw an unsupported error.

**Comment by ddorian** [comment]
@dbotwinick the tables & indexes are clustered on disk by default https://docs.yugabyte.com/latest/architecture/docdb/ by the PRIMARY KEY.

And I don't think there's a chance in the near/medium term to support clustering on something else besides the primary key. So for now you can just ignore this error. 

We'll look internally into enabling it and making it a no-op for now.

**Comment by dbotwinick** [comment]
> @dbotwinick the tables & indexes are clustered on disk by default https://docs.yugabyte.com/latest/architecture/docdb/ by the PRIMARY KEY.
> 
> And I don't think there's a chance in the near/medium term to support clustering on something else besides the primary key. So for now you can just ignore this error.
> 
> We'll look internally into enabling it and making it a no-op for now.

@ddorian that would work for me! I haven't done particular testing on the performance implications that could possibly come up here... but my gut feeling is that it wouldn't particularly be a problem in this case. Ideally I could just use the vanilla application without modification, so if it were a no-op and it logs a warning or something--that would be sufficient for my purposes. It would also work for me if there were some sort of configuration flags to switch between no-op while logging issue and throwing an error. I don't know how treating it as a no-op might cause trouble for others, so it might be safer to make it a configurable option so that the defaults are safer? (although that comes with its own issues)

**Comment by pcmid** [comment]
+1 I can't migrate from my dump

ERROR:  ALTER TABLE ALTER column not supported yet
LINE 1: ALTER TABLE ONLY public.event_search ALTER COLUMN room_id SE...

```sql
CREATE TABLE public.event_search (
    event_id text,
    room_id text,
    sender text,
    key text,
    vector tsvector,
    origin_server_ts bigint,
    stream_ordering bigint
);
ALTER TABLE ONLY public.event_search ALTER COLUMN room_id SET (n_distinct=-0.01);
```

**Comment by ddorian** [comment]
Hi @pcmid 

> ALTER TABLE ONLY public.event_search ALTER COLUMN room_id SET (n_distinct=-0.01);

Can you explain the use case behind manually changing the statistics? Do you know in advance the number of unique values in the column?

**Comment by pcmid** [comment]
> Hi @pcmid
> 
> > ALTER TABLE ONLY public.event_search ALTER COLUMN room_id SET (n_distinct=-0.01);
> 
> Can you explain the use case behind manually changing the statistics? Do you know in advance the number of unique values in the column?

Thanks for reply. The reason is [here](https://github.com/matrix-org/synapse/blob/7d3ea4886c09c3c3557a876521ed99e58f6f13f9/synapse/storage/schema/main/delta/73/11event_search_room_id_n_distinct.sql.postgres#L17-L28).
Maybe this is only needed in postgresql?

**Comment by purvish-bs** [comment]
> @dodilp - FOREIGN KEY is being implemented as we speak. Hope to have it in a few weeks.

Is this implemented?

**Comment by ddorian** [comment]
@purvish-bs yes https://docs.yugabyte.com/preview/explore/indexes-constraints/foreign-key-ysql/

**Comment by purvish-bs** [comment]
I am not a pro in SQL, Could anyone please explain what is not supported here? Thanks in Advance

![image](https://github.com/yugabyte/yugabyte-db/assets/127775625/9e7de2df-bd3e-421c-9175-3d11f603bbbb)

**Comment by ddorian** [comment]
@purvish-bs 

can you provide the full sql migration?
I have to look at your schema. I think you may have linked the table to a composite type, not really sure though without having all the migration.

**Comment by fizaaluthra** [comment]
Known issues for ALTER TYPE:
https://github.com/yugabyte/yugabyte-db/issues/17756 (ALTER TABLE ALTER COLUMN TYPE fails when on range key table with split options)
https://github.com/yugabyte/yugabyte-db/issues/18066 (ALTER TYPE fails after DROP COLUMN)

**Comment by ctr49** [comment]
I'm seeing this in a real world scenario upon schema installation, see https://github.com/element-hq/synapse/blob/master/synapse/storage/schema/state/full_schemas/72/full.sql.postgres#L23

Are there any workarounds available on Yugabyte side? I could obviously change the schema to make it compatible, but then I couldn't rely on Synapse schema upgrades going forward...

**Comment by ddorian** [comment]
@ctr49 

> Are there any workarounds available on Yugabyte side? I could obviously change the schema to make it compatible, but then I couldn't rely on Synapse schema upgrades going forward...

There are no current workarounds. You can make a fork of synapse and just comment that part and you still can rely on them for the other upgrades. You can also subscribe to issue https://github.com/yugabyte/yugabyte-db/issues/16675 for updates.

**Comment by lucashamaguchi** [comment]
+1 for `ALTER TABLE name SET (storage_option = value)`

```sql
alter table column set (fillfactor = 75);
```

**Comment by OlegLoginov** [comment]
Related:    https://github.com/yugabyte/yugabyte-db/issues/23889

**Comment by ddorian** [comment]
Hi @tz-torchai 

Can you specify the 3rd party app if it's open source?
